File: RQ3/Java-REval-Desc/task46_desc.java
public class Translator {
    public static void byLength(int[] arr) {
        java.util.Map<Integer, String> dic = new java.util.HashMap<>();
        dic.put(1, "One");
        dic.put(2, "Two");
        dic.put(3, "Three");
        dic.put(4, "Four");
        dic.put(5, "Five");
        dic.put(6, "Six");
        dic.put(7, "Seven");
        dic.put(8, "Eight");
        dic.put(9, "Nine");
        
        java.util.Arrays.sort(arr);
        java.util.List<String> newArr = new java.util.ArrayList<>();
        for (int i = arr.length - 1; i >= 0; i--) {
            if (dic.containsKey(arr[i])) {
                newArr.add(dic.get(arr[i]));
            }
        }
        String[] result = newArr.toArray(new String[0]);

        assert func(result); // POST: func is equivalent to: Given an array of integers, sort the integers that are between 1 and 9 inclusive, reverse the resulting array, and then replace each digit by its corresponding name "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine". For example: arr = [2, 1, 1, 4, 5, 8, 2, 3] -> sort arr -> [1, 1, 2, 2, 3, 4, 5, 8] -> reverse arr -> [8, 5, 4, 3, 2, 2, 1, 1] return ["Eight", "Five", "Four", "Three", "Two", "Two", "One", "One"] If the array is empty, return an empty array: arr = [] return [] If the array has any strange number ignore it: arr = [1, -1 , 55] -> sort arr -> [-1, 1, 55] -> reverse arr -> [55, 1, -1] return = ['One']
    }
}

File: RQ3/C-REval-Desc/task46_desc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char* dic[] = {
    "Zero", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine"
};

void by_length(int* arr, int size, const char*** result, int* result_size) {
    int* sorted_arr = (int*)malloc(size * sizeof(int));
    int count = 0;

    for (int i = 0; i < size; i++) {
        if (arr[i] >= 1 && arr[i] <= 9) {
            sorted_arr[count++] = arr[i];
        }
    }

    qsort(sorted_arr, count, sizeof(int), (int (*)(const void*, const void*))cmp);

    char** new_arr = (char**)malloc(count * sizeof(char*));
    for (int i = count - 1; i >= 0; i--) {
        new_arr[count - 1 - i] = dic[sorted_arr[i]];
    }

    *result = new_arr;
    *result_size = count;

    free(sorted_arr);
}

int cmp(const int* a, const int* b) {
    return (*a > *b) - (*a < *b);
}

int func(const char** arr, int size); // Forward declaration for the assertion

int main() {
    int arr[] = {2, 1, 1, 4, 5, 8, 2, 3};
    const char** result;
    int result_size;

    by_length(arr, sizeof(arr) / sizeof(arr[0]), &result, &result_size);

    assert(func(result, result_size));

    free(result);
    return 0;
}